feat: add blog tag browsing for course sections [SCHOOL-58, SCHOOL-61]

ryon opened the pull request:
Related: #236 

- Upgrade FontAwesome from 4.0.3 to to 4.7.0 (latest minor version)
- Use wells for all class sidebar items
- Improve styling of class sidebar
- Add mockup for blog tag browser

![Screen Shot 2021-03-09 at 11 48 06 PM](https://user-images.githubusercontent.com/1154929/110578212-4b809d00-8132-11eb-88c2-8a6e6dff92d6.png)

![Screen Shot 2021-03-09 at 11 48 13 PM](https://user-images.githubusercontent.com/1154929/110578214-4b809d00-8132-11eb-8841-eb0a1ce55c0b.png)

ryon committed:
feat: improve course section sidebar styling and add tag browser mockup

BillClinton committed:
feat: update Section obj to provide posts and tags

BillClinton committed:
fix: add BlogPost class to imports

BillClinton committed:
fix: correct tag counting

BillClinton committed:
feat: implement tag links in section template

BillClinton committed:
fix: improve function name

BillClinton committed:
feat: move code from template to Class function

BillClinton committed:
feat: update count variable to reflect new name

BillClinton committed:
feat: update func to use TagItem::getTagsSummary

BillClinton committed:
feat: add browsing section blog posts by tag

BillClinton committed:
feat: add fixture data to test tag browsing

BillClinton committed:
feat: Don't pin teacher post when tag browsing

BillClinton committed:
fix: remove test code

BillClinton committed:
fix: clean up fixture files

BillClinton committed:
temp: Override emergence-skeleton class

This overide matches the changes made in https://github.com/JarvusInnovations/emergence-skeleton/pull/318
and the file can be removed when upstream PR is merged

BillClinton reviewed (commented):
TODO: Before merging, delete this file and merge upstream PR: https://github.com/JarvusInnovations/emergence-skeleton/pull/318

BillClinton committed:
feat: move blog/tag logic to request handler

Commit:
fix: correct blog template path

BillClinton committed:
fix: add history tables to new fixtures

BillClinton committed:
temp: fix: allow for empty subquery result

This file is a placeholder override for changes on emergence-skeleton.
It can be deleted when merged with updated emergence-skeleton.

BillClinton committed:
fix: allow for zero results from subquery

BillClinton committed:
feat: hide blog tag box when empty and update name

BillClinton committed:
fix: send current blog tag with data for header

BillClinton committed:
temp: fix: allow for empty subquery result

This file is a placeholder override for changes on emergence-skeleton.
It can be deleted when merged with updated emergence-skeleton.

themightychris commented:
@BillClinton your upstream changes are now merged into `slate`'s `develop` branch via 17397fe02793f90e5096338d1a16768dcfad15fc

If you `git fetch && git rebase origin/develop` this branch you can then delete the TagItem overrides

BillClinton committed:
fix: changes have now been added to slate-skeleton

This file was acting as an override.  The changes made to it
have now been incorporated into the slate-skeleton version and
this file can now be removed.

themightychris commented on html-templates/sections/courseSection.tpl:
```diff
@@ -157,27 +153,91 @@
+                        <ul class="tag-list">
+                        {foreach item=tag from=$tagsShown}
+                            <li>
+                                <a class="tag" href="?blog_tag={$tag.Handle}">{$tag.Title} <span class="tag-count">{$tag.itemsCount}</span></a>
```
```suggestion
                                <a class="tag" href="?blog_tag={$tag.Handle}">{$tag.Title|escape} <span class="tag-count">{$tag.itemsCount}</span></a>
```

themightychris commented on html-templates/sections/courseSection.tpl:
```diff
@@ -157,27 +153,91 @@
+                                <ul class="tag-list">
+                                {foreach item=tag from=$tagsHidden}
+                                    <li>
+                                        <a class="tag" href="?blog_tag={$tag.Handle}">{$tag.Title} <span class="tag-count">{$tag.count}</span></a>
```
```suggestion
                                        <a class="tag" href="?blog_tag={$tag.Handle}">{$tag.Title|escape} <span class="tag-count">{$tag.count}</span></a>
```

themightychris commented on php-classes/Slate/Courses/SectionsRequestHandler.php:
```diff
@@ -110,11 +111,52 @@ public static function handleRecordRequest(ActiveRecord $Section, $action = fals
+
+      $limit = isset($_REQUEST['limit']) && ctype_digit($_REQUEST['limit']) ? (integer)$_REQUEST['limit'] : 10;
+      $offset = isset($_REQUEST['offset']) && ctype_digit($_REQUEST['offset']) ? (integer)$_REQUEST['offset'] : 0;
+      $handle = $_REQUEST['blog_tag'];
```
use `blog-tag` -- lowerhyphencase is preferred for query params by the internet at large

themightychris replied:
@BillClinton doesn't look like this was resolved

Edit: looks like we have `blog-tag` in other spots and the filter currently doesn't work on this PR because of this mismatch

themightychris replied:
oh looks like you changed it in other places but not here, so I'd expect it to not be working as-is

themightychris commented on php-classes/Slate/Courses/SectionsRequestHandler.php:
```diff
@@ -110,11 +111,52 @@ public static function handleRecordRequest(ActiveRecord $Section, $action = fals
                 ]));
             case 'students':
                 return static::handleStudentsRequest($Section);
+            case false:
```
```suggestion
            case '':
            case false:
```

As can be found in the underlying `handleRecordRequest` implementation, we need both cases to catch `/blog/myblog` and `/blog/myblog/`

themightychris commented on php-classes/Slate/Courses/SectionsRequestHandler.php:
```diff
@@ -110,11 +111,52 @@ public static function handleRecordRequest(ActiveRecord $Section, $action = fals
         }
     }
 
+    public static function handleBlogpostRequest(Section $Section, $action)
```
```suggestion
    public static function handleBlogpostRequest(Section $Section, $action)
```

For one, we'd want to call this `handleBlogPostRequest` to match how BlogPost is usually cased in Slate, but also this is `handleSectionRequest` -- the primary response payload and thing being requested is still a section, we just may or may not include related blog post data

themightychris commented on php-classes/Slate/Courses/SectionsRequestHandler.php:
```diff
@@ -110,11 +111,52 @@ public static function handleRecordRequest(ActiveRecord $Section, $action = fals
+      $tag = null;
+
+      if ($handle) {
+          $tag = Tag::getByHandle($handle);
```
convention in emergence is to capitalize `$Tag` when it is holding an ActiveRecord instance

themightychris commented on php-classes/Slate/Courses/SectionsRequestHandler.php:
```diff
@@ -110,11 +111,52 @@ public static function handleRecordRequest(ActiveRecord $Section, $action = fals
+          $tag = Tag::getByHandle($handle);
+
+          if (!$tag) {
+              static::throwNotFoundError('tag not found');
```
```suggestion
              return static::throwNotFoundError('tag not found');
```

response calls need to be returned to ensure the function exits and because some response modes return data up the call chain

themightychris commented on php-classes/Slate/Courses/Section.php:
```diff
@@ -411,6 +419,68 @@ public function getCohorts()
+
+      if ($tag!=null) {
+
+          $tagIDsQuery = 'SELECT ContextID FROM `tag_items` WHERE (`ContextClass` = "%s") AND (`TagID` = %u)';
+          $tagIDsConditions = [
+              'ContextClass' => DB::escape(BlogPost::getStaticRootClass()),
+              'TagID' => $tag->ID
+          ];
+
+          $tagItemIDs = DB::allValues('ContextID', $tagIDsQuery, $tagIDsConditions);
```
`$conditions` and the `'Field' => 'value'` syntax is commonly used when passing a conditions structure to high-level ActiveRecord methods that parse that structure to generate SQL -- most importantly escaping is handled automatically in these cases

In this case, low-level `DB` methods don't have any of those smarts, they are just wrapping around `sprintf`. The use of the name `conditions` is misleading, and using named keys is misleading because they are ignored and this only works because of the position of the two args

I would prefer this form:

```suggestion
          $tagItemIDs = DB::allValues(
              'ContextID',
              'SELECT ContextID FROM `tag_items` WHERE (`ContextClass` = "%s") AND (`TagID` = %u)',
              [
                  DB::escape(BlogPost::getStaticRootClass(),
                  $tag->ID
              ]
          );
```

As it makes the positional connection clearer

themightychris commented on php-classes/Slate/Courses/Section.php:
```diff
@@ -411,6 +419,68 @@ public function getCohorts()
+          return $Teacher->ID;
+      }, $this->ActiveTeachers);
+
+      $latestTeacherPost = \Emergence\CMS\BlogPost::getAllPublishedByContextObject($this, array_merge_recursive([
```
put a `use` statement up top and just do `BlogPost`

themightychris commented on php-classes/Slate/Courses/Section.php:
```diff
@@ -411,6 +419,68 @@ public function getCohorts()
+          'limit' => 1
+      ]));
+
+      return $latestTeacherPost[0];
```
to avoid undefined index error if there are no results:

```suggestion
      return count($latestTeacherPost) ? $latestTeacherPost[0] : null;
```

themightychris reviewed (changes requested):
You found the perfect way to implement everything! Well done!

Just some minor style/architecture tweaks to implement and then we're good to go

BillClinton committed:
fix: escape text output

BillClinton committed:
fix: use lowerhyphencase for query param

BillClinton committed:
fix: Error should be returned

BillClinton committed:
fix: avoid undefined index error if no results

BillClinton committed:
fix: no need to specify full path of class

BillClinton committed:
fix: catch both cases

BillClinton committed:
fix: use more appropriate method name

BillClinton committed:
refactor: remove unnecessarily named args

themightychris reviewed (changes requested):
almost there!

BillClinton committed:
fix: corrected url parameter name

themightychris committed:
style(blog): wrap tag in <code> tag
